### PR TITLE
Support generateMetadata export

### PR DESCRIPTION
**Pull request by huozhi**
Closes NEXT-397

Resolve `metadata` and `generateMetadata()` exports along with head during rendering, this is the easy way for now to collect all the metadata properly. Since we can access segment params and search params only in rendering, so I moved all the resolving logic from loader to render process.

## Feature

- [x] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [x] [e2e](https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs) tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 1929f5f00db2406e9384ad18ec8cde6103e2be43

`pnpm testheadless test/e2e/app-dir/metadata/metadata.test.ts` 
- app dir - metadata > metadata > basic > should apply metadata when navigating client-side

<details>
<summary>Expand output</summary>

  ● app dir - metadata › metadata › basic › should apply metadata when navigating client-side

    expect(received).toBe(expected) // Object.is equality

    Expected: "origin-when-crossorigin"
    Received: undefined

      47 |           expect(values).toEqual(content)
      48 |         } else {
    > 49 |           expect(values[0]).toBe(content)
         |                             ^
      50 |         }
      51 |       }
      52 |

      at checkMeta (e2e/app-dir/metadata/metadata.test.ts:49:29)
      at Object.<anonymous> (e2e/app-dir/metadata/metadata.test.ts:281:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/e2e/app-dir/metadata/app/apple/page.tsx [hunk 1]
```diff
@@ -2,20 +2,23 @@ export default function page() {
...
+      startupImage: [
+        '/assets/startup/apple-touch-startup-image-768x1004.png',
+        {
+          url: '/assets/startup/apple-touch-startup-image-1536x2008.png',
+          media: '(device-width: 768px) and (device-height: 1024px)',
+        },
+      ],
+    },
+  }
+  return metadata
```
Comment (shuding) [inline test/e2e/app-dir/metadata/app/apple/page.tsx:23]: Can you add tests to confirm that it receives these 2 arguments?

**Review by shuding — DISMISSED** [review]

**Review by timneutkens — CHANGES_REQUESTED** [review]
This doesn't handle the cache() deduping yet it seems. Will double check first.

**Review by timneutkens — CHANGES_REQUESTED** [review]
Checked and added some tests that are failing 👍

**Review by timneutkens — APPROVED** [review]

**Pull request merged**